### PR TITLE
Add HMR support,  index file support, log errors from server

### PR DESCRIPTION
**Pull request by maxaon**
Add HMR support, index file generating support, log errors from server to stderr

### CODE REVIEW

File: config/project.js [hunk 1]
```diff
@@ -33,7 +33,8 @@ module.exports = {
   },
   separateCss: () => getConfig('separateCss', true),
   cssModules: () => getConfig('cssModules'),
-  externals: () => getConfig('externals')
+  externals: () => getConfig('externals'),
+  indexFile: () => getConfig('index')
```
Comment (mastertheblaster) [inline config/project.js]: FYI: We could have multiple index files.

File: test/start.spec.js [hunk 2]
```diff
@@ -1,32 +1,48 @@
 'use strict';
 
-const _ = require('lodash/fp');
+const fp = require('lodash/fp');
```
Comment (bildja) [inline test/start.spec.js]: Please don't name it `fp`

**Comment by shahata** [comment]
@maxaon plz check why build in pullrequest teamcity is failing

File: config/webpack.config.client.js [hunk 3]
```diff
@@ -50,6 +59,33 @@ const config = ({debug} = {}) => {
       return new ExtractTextPlugin(debug ? '[name].css' : '[name].min.css');
     }
   }
+
+  function addHotEntries(entries) {
+    return _.mapValues(entries, entry => {
+      entry = _.isArray(entry) ? entry : [entry];
+      return [
+        require.resolve('webpack-hot-middleware/client') + '?reload=true',
```
Comment (ronami) [inline config/webpack.config.client.js]: HMR is not working, the Webpack server is running on port 3200 while webpack-hot-middleware/client tries to connect to port 3000.

File: config/webpack.config.client.js [hunk 4]
```diff
@@ -50,6 +59,33 @@ const config = ({debug} = {}) => {
...
+  function getIndexPlugin(entries, options) {
+    if (_.isString(options)) {
+      options = {template: options};
+    }
+    options = options || {};
+    const minSuffix = debug ? '' : '.min';
+    return Object.keys(entries)
+      .map(entryName => {
+        const substitute = tpl => tpl && tpl.replace(/\[name]/g, entryName);
+        return new HtmlWebpackPlugin(_.pickBy({ //  plugin uses _.extend, so we need to provide only non empty values
```
Comment (ronami) [inline config/webpack.config.client.js]: Why do we need HtmlWebpackPlugin? Is it related to HMR or is it a separate feature?

File: config/webpack.config.common.js [hunk 5]
```diff
@@ -25,7 +25,8 @@ const config = {
       require('../lib/loaders/babel')(),
       require('../lib/loaders/typescript')(),
       require('../lib/loaders/images')(),
-      require('../lib/loaders/json')()
+      require('../lib/loaders/json')(),
+      require('../lib/loaders/html')()
```
Comment (ronami) [inline config/webpack.config.common.js]: Why do we need html-loader? Is it related to HMR or is it a separate feature?

File: package.json [hunk 6]
```diff
@@ -72,20 +73,23 @@
     "mocha-env-reporter": "^1.0.2",
     "mocha-teamcity-reporter": "^1.1.1",
     "node-sass": "~3.9.0",
+    "npm": "^3.10.8",
```
Comment (ronami) [inline package.json]: Why do we need a dependency on npm?

File: wix-node-build-start.js [hunk 7]
```diff
@@ -11,6 +11,7 @@ program
   .option('-n, --no-server', 'run without starting entry-point')
   .option('-w, --watch', 'deprecated, for backward comp.')
   .option('-s, --silent', 'deprecated, for backward comp.')
+  .option('-h, --hot', 'use hot module replacement')
```
Comment (ronami) [inline wix-node-build-start.js]: Is there any reason to put HMR under a flag? @shahata can we have it without a flag, or at least on by default?

**Comment by ronami** [comment]
Closing PR for now.

I merged in HMR support (https://github.com/wix/wix-node-build/commit/3bba4cf4542511b44d6eb03a8e515cee561a79e0) and it was decided to only show a brief message to let users know if there are errors, while the error call-stack is still in target/server.log (https://github.com/wix/wix-node-build/commit/47a4d8afcb0edad98b3ce9b5e7d0a364d0423b41).

As I can see PR also includes: code refactor, support for html-loader, and the use of HtmlWebpackPlugin. Please open new PRs if any of them is still relevant. If you need my help with anything please let me know.

**Pull request closed without merging**